pkg/nsx/services/ippool/ippool.go

```diff
@@ -0,0 +1,252 @@
+
+	wg.Add(2)
+
+	ipPoolService := &IPPoolService{Service: service}
+	ipPoolService.ipPoolStore = &IPPoolStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeIPPoolCRUID: indexFunc}),
+		BindingType: model.IpAddressPoolBindingType(),
+	}}
+	ipPoolService.ipPoolBlockSubnetStore = &IPPoolBlockSubnetStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeIPPoolCRUID: indexFunc}),
+		BindingType: model.IpAddressPoolBlockSubnetBindingType(),
+	}}
+
+	go ipPoolService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeIPPool, ipPoolService.ipPoolStore)
+	go ipPoolService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeIPPoolBlockSubnet, ipPoolService.ipPoolBlockSubnetStore)
+
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return ipPoolService, err
+	}
+	return ipPoolService, nil
+}
+
+func (service *IPPoolService) CreateOrUpdateIPPool(obj *v1alpha2.IPPool) (bool, bool, error) {
+	nsxIPPool, nsxIPSubnets := service.BuildIPPool(obj)
+	if nsxIPPool == nil || len(nsxIPSubnets) == 0 {
+		err := util.NoEffectiveOption{Desc: "build ip pool and ip pool subnets failed, check its namespace, ippool type and vpc"}
+		return false, false, err
+	}
+	existingIPPool, existingIPSubnets, err := service.indexedIPPoolAndIPPoolSubnets(obj.UID)
+	log.V(1).Info("existing ippool and ip subnets", "existingIPPool", existingIPPool, "existingIPSubnets", existingIPSubnets)
+	if err != nil {
+		log.Error(err, "failed to get ip pool and ip pool subnets by UID", "UID", obj.UID)
+		return false, false, err
+	}
+	ipPoolSubnetsUpdated := false
+	ipPoolUpdated := common.CompareResource(IpAddressPoolToComparable(existingIPPool), IpAddressPoolToComparable(nsxIPPool))
+	changed, stale := common.CompareResources(IpAddressPoolBlockSubnetsToComparable(existingIPSubnets), IpAddressPoolBlockSubnetsToComparable(nsxIPSubnets))
+	changedIPSubnets, staleIPSubnets := ComparableToIpAddressPoolBlockSubnets(changed), ComparableToIpAddressPoolBlockSubnets(stale)
+	for i := len(staleIPSubnets) - 1; i >= 0; i-- {
+		staleIPSubnets[i].MarkedForDelete = &MarkedForDelete
+	}
+	finalIPSubnets := append(changedIPSubnets, staleIPSubnets...)
+	if len(finalIPSubnets) > 0 {
+		ipPoolSubnetsUpdated = true
+	}
+
+	if err := service.Operate(nsxIPPool, finalIPSubnets, ipPoolUpdated, ipPoolSubnetsUpdated); err != nil {
+		return false, false, err
+	}
+
+	realizedSubnets, subnetCidrUpdated, e := service.AcquireRealizedSubnetIP(obj)
+	if e != nil {
+		return false, false, e
+	}
+	obj.Status.Subnets = realizedSubnets
+	return subnetCidrUpdated, ipPoolSubnetsUpdated, nil
+}
+
+func (service *IPPoolService) Operate(nsxIPPool *model.IpAddressPool, nsxIPSubnets []*model.IpAddressPoolBlockSubnet, IPPoolUpdated bool, IPPoolSubnetsUpdated bool) error {
+	if !IPPoolUpdated && !IPPoolSubnetsUpdated {
```

Comment:
can we change to if !(IPPoolUpdated  || IPPoolSubnetsUpdated )?

Reply:
ok